contracts/cw1-subkeys/src/contract.rs

```diff
@@ -251,14 +252,142 @@ mod tests {
 
     #[test]
     fn query_allowances() {
-        // TODO
-        // check the allowances work for accounts with balances and accounts with none
+        let mut deps = mock_dependencies(20, &coins(1111, "token1"));
+
+        let owner = HumanAddr::from("admin0001");
+        let admins = vec![owner.clone(), HumanAddr::from("admin0002")];
+
+        let spender1 = HumanAddr::from("spender0001");
+        let spender2 = HumanAddr::from("spender0002");
+        let spender3 = HumanAddr::from("spender0003");
+        let initial_spenders = vec![spender1.clone(), spender2.clone()];
+
+        // Same allowances for all spenders, for simplicity
+        let denom1 = "token1";
+        let amount1 = 1111;
+
+        let allow1 = coin(amount1, denom1);
+        let initial_allowances = vec![allow1.clone()];
+
+        let expires_never = Expiration::Never {};
+        let initial_expirations = vec![expires_never.clone(), expires_never.clone()];
+
+        let env = mock_env(owner, &[]);
+        setup_test_case(
+            &mut deps,
+            &env,
+            &admins,
+            &initial_spenders,
+            &initial_allowances,
+            &initial_expirations,
+        );
+
+        // Check allowances work for accounts with balances
+        let allowance = query_allowance(&deps, spender1.clone()).unwrap();
+        assert_eq!(
+            allowance,
+            Allowance {
+                balance: Balance(vec![allow1.clone()]),
+                expires: expires_never.clone()
+            }
+        );
+        let allowance = query_allowance(&deps, spender2.clone()).unwrap();
+        assert_eq!(
+            allowance,
+            Allowance {
+                balance: Balance(vec![allow1.clone()]),
+                expires: expires_never.clone()
+            }
+        );
+
+        // Check allowances work for accounts with no balance
+        let allowance = query_allowance(&deps, spender3.clone()).unwrap();
+        assert_eq!(allowance, Allowance::default(),);
     }
 
     #[test]
     fn update_admins_and_query() {
-        // TODO
-        // insure imported logic is wired up properly
+        let mut deps = mock_dependencies(20, &coins(1111, "token1"));
+
+        let owner = HumanAddr::from("admin0001");
+        let admin2 = HumanAddr::from("admin0002");
+        let admin3 = HumanAddr::from("admin0003");
+        let initial_admins = vec![owner.clone(), admin2];
+
+        let env = mock_env(owner.clone(), &[]);
+        setup_test_case(&mut deps, &env, &initial_admins, &vec![], &vec![], &vec![]);
+
+        // Verify
+        let config = query_config(&deps).unwrap();
+        assert_eq!(
+            config,
+            ConfigResponse {
+                admins: initial_admins.clone(),
+                mutable: true,
+            }
+        );
+
+        // Add a third (new) admin
+        let msg = HandleMsg::UpdateAdmins {
+            admins: [&config.admins[..], &[admin3.clone()]].concat(),
```

Comment:
Nice use of the spread and concat operators. But somehow, I would prefer to see this more explicit.

```rust
let new_admins = vec![owner.clone(), admin2.clone(), admin3.clone()]
let msg = HandleMsg::UpdateAdmins { admins: new_admins.clone() };
// ...
assert_eq!(config, ConfigResponse { admins: new_admins, mutable: true });
```

This is the exact same functionality you have written but easier to read. And key in test cases is legibility (performance is not an issue)